File: 20-valid-parentheses/valid-parentheses.cpp
#define MAXI 20000
class Solution {
public:
    char stack[MAXI];
    int pointer = 0;

    void push(char ch) {
        if (pointer < MAXI) {pointer++;
            stack[pointer] = ch;
            
        } else
            return;
    }
    char pop() {
        char ch;
        if (pointer > 0) {
            ch = stack[pointer];
            pointer--;
        }
         

        return ch;
    }
    bool isValid(string s) {
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
                push(s[i]);
            }
            
            if (!(s[i] == '(' || s[i] == '{' || s[i] == '[')) {
                if(pointer==0)
            {
                return false;
            }
                char ch = pop();
                if ((ch == '(' && s[i] != ')') || (ch == '{' && s[i] != '}') ||
                    (ch == '[' && s[i] != ']')) { // the false need to verified only once so
                                     // we can use this for our program.
                    return false;
                }
            }
        }
        return pointer==0;
    }
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack=[]
        open=['[','(','{']
        pair={"[":"]","(":")","{":"}"}
        if len(s)==1:
            return False
        for i in s:
            if i in open :
                stack.append(i)
            elif stack:
                 s=stack.pop()
                 if pair[s]!=i:
                    return False
            else:
                return False
        return True if len(stack) == 0 else False